runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerWrapper.java

```diff
@@ -0,0 +1,127 @@
+class DockerWrapper {
+  // TODO: Should we require 64-character container ids? Docker technically allows abbreviated ids,
+  // but we _should_ always capture full ids.
+  private static final Pattern CONTAINER_ID_PATTERN = Pattern.compile("\\p{XDigit}{64}");
+
+  static DockerWrapper forCommand(String dockerExecutable, Duration commandTimeout) {
+    return new DockerWrapper(dockerExecutable, commandTimeout);
+  }
+
+  private final String dockerExecutable;
+  private final Duration commandTimeout;
+
+  private DockerWrapper(String dockerExecutable, Duration commandTimeout) {
+    this.dockerExecutable = dockerExecutable;
+    this.commandTimeout = commandTimeout;
+  }
+
+  /**
+   * Runs the given container image with the given command line arguments. Returns the running
+   * container id.
+   */
+  public String runImage(String imageTag, List<String> args)
+      throws IOException, TimeoutException, InterruptedException {
+    checkArgument(!imageTag.isEmpty(), "Docker image tag required");
+    // TODO: Validate args?
+    return runShortCommand(
+        ImmutableList.<String>builder()
+            .add(dockerExecutable)
+            .add("run")
+            .add("-d")
+            .add(imageTag)
+            .addAll(args)
+            .build());
+  }
+
+  /**
+   * Kills a docker container by container id.
+   *
+   * @throws IOException if an IOException occurs or if the given container id does not exist
+   */
+  public void killContainer(String containerId)
+      throws IOException, TimeoutException, InterruptedException {
+    checkArgument(containerId != null);
+    checkArgument(
+        CONTAINER_ID_PATTERN.matcher(containerId).matches(),
+        "Container ID must be a 64-character hexadecimal string");
+    runShortCommand(Arrays.asList(dockerExecutable, "kill", containerId));
+  }
+
+  /** Run the given command invocation and return stdout as a String. */
+  private String runShortCommand(List<String> invocation)
+      throws IOException, TimeoutException, InterruptedException {
+    ProcessBuilder pb = new ProcessBuilder(invocation);
+    Process process = pb.start();
+    // TODO: Consider supplying executor service here.
+    CompletableFuture<String> resultString =
+        CompletableFuture.supplyAsync(
+            () -> {
+              // NOTE: We do not own the underlying stream and do not close it.
+              BufferedReader reader =
+                  new BufferedReader(
+                      new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
+              return reader.lines().collect(Collectors.joining());
+            });
+    // TODO: Retry on interrupt?
+    boolean processDone = process.waitFor(commandTimeout.toMillis(), TimeUnit.MILLISECONDS);
+    if (!processDone) {
+      process.destroy();
+      throw new TimeoutException(
+          String.format(
+              "Timed out while waiting for command '%s'",
```

Comment:
<!--new_thread; commit:eb0cdf4ea2d094d8ab09a0a9ebd742507f71a1a8; resolved:0-->
In this and the other error case, is it possible to log the output from the command? That'd be pretty helpful for debugging why it failed.